File: js/examples/gapAudio.ts
/**
 * Two-party GAP (audio) frame exchange — minimal in-memory end-to-end example.
 *
 * Demonstrates:
 *   - Sending synthetic Opus frames with FlatBuffers codec (recommended for audio)
 *   - Sequential frames: rtp_sequence advances automatically inside GapClient
 *
 * Run from repo root:
 *   npx ts-node js/examples/gapAudio.ts
 */

import { MlsContext } from "../src/mls";
import { GroupNode, StreamType } from "../src/node";
import { GapClient } from "../src/gap";
import { PayloadCodec } from "../src/native";

const aliceMls = MlsContext.create("alice");
const bobMls   = MlsContext.create("bob");
bobMls.acceptWelcome(aliceMls.invite(bobMls.exportKeyPackage()));

const gid   = aliceMls.groupId;
const alice = GroupNode.create(1, gid);
const bob   = GroupNode.create(2, gid);
alice.bootstrapAsCreator(aliceMls.epoch);
bob.bootstrapAsJoiner(bobMls.epoch);

const gapAlice = GapClient.create();
const gapBob   = GapClient.create();

// Synthetic 20 ms Opus frame (zeroed; real usage: encode from PCM).
const opus = Buffer.alloc(40);

for (let i = 0; i < 3; i++) {
    const frame = gapAlice.send(
        alice, aliceMls,
        /*target*/          2,
        /*mediaSourceId*/   1,
        /*rtpTimestamp*/    BigInt(i * 960),
        opus,
        PayloadCodec.FlatBuffers,  // lowest decode latency
    );
    for (const ev of bob.onWire(bobMls, frame.wire)) {
        if (ev.kind === "payload_received" && ev.streamType === StreamType.Audio) {
            const r = gapBob.accept(ev.plaintext!, bobMls.epoch, ev.codec ?? PayloadCodec.Cbor);
            console.log(`frame ${i + 1}: status=${r.status}  seq=${r.seq}  codec=${ev.codec}`);
        }
    }
}

[aliceMls, bobMls, alice, bob, gapAlice, gapBob].forEach(x => x.close());

File: python/examples/gap_audio.py
"""Two-party GAP (audio) frame exchange — minimal in-memory end-to-end example.

Demonstrates:
  - Sending synthetic Opus frames with FlatBuffers codec (recommended for audio)
  - Sequential frames: rtp_sequence advances automatically inside GapClient

Run:
    python python/examples/gap_audio.py
"""

from gbp_stack import MlsContext, GroupNode, GapClient, PayloadCodec

with MlsContext.create("alice") as alice_mls, \
     MlsContext.create("bob")   as bob_mls:

    bob_mls.accept_welcome(alice_mls.invite(bob_mls.export_key_package()))

    gid = alice_mls.group_id
    with GroupNode.create(member_id=1, group_id=gid) as alice, \
         GroupNode.create(member_id=2, group_id=gid) as bob, \
         GapClient.create() as gap_alice, \
         GapClient.create() as gap_bob:

        alice.bootstrap_as_creator(alice_mls.epoch)
        bob.bootstrap_as_joiner(bob_mls.epoch)

        # Synthetic 20 ms Opus frame (zeroed; real usage: encode from PCM).
        opus = bytes(40)

        for i in range(3):
            frame = gap_alice.send(
                alice, alice_mls,
                target=2,
                media_source_id=1,
                rtp_timestamp=i * 960,
                opus=opus,
                codec=PayloadCodec.FLATBUFFERS,  # lowest decode latency
            )
            for ev in bob.on_wire(bob_mls, frame.wire):
                if ev.kind == "payload_received" and ev.stream_type == 1:
                    r = gap_bob.accept(ev.plaintext or b"", bob_mls.epoch,
                                      codec=ev.codec or PayloadCodec.CBOR)
                    print(f"frame {i+1}: status={r.status}  seq={r.seq}  codec={ev.codec}")
